Consistently handle the 'disabled' attr in inputs

abeluck opened the pull request:
As discussed in https://community.zammad.org/t/automating-creation-of-custom-object-attributes/3831/5

This commit ensures that the generic input widgets handle the `disabled`
attribute consistently. Before only the checkbox, select and input widgets respected
this attribute.

I've added it to the following widgets

* tree/searchable select
* textarea
* datetime
* date

Now it is possible to define a screens policy like so:
```
screens: {
"edit"=>{
"ticket.customer"=>{"shown"=>true, "required"=>false},
"ticket.agent"=>{
  "shown"=>true,
  "required"=>false,
  "disabled"=>true # <--- add this
}}}
```

Then the fields will show up on the Agent's sidebar with the customer
submitted values, but the Agent will not be able to change the values.

This is not the same as real permissions, as presumably a manually
crafted API request could circumvent the disabled attribute.

But this is a first step towards that direction, and very useful for us
in prod right now.

<!--
Hi there - a lot of love for starting a Pull Request 😍. Please ensure the following things before creation - thank you!

- Create your Pull Request against the develop branch. We don't accept changes to stable branches.
- Add a reference to the issue you are trying to fix. Just add a # and the number.
- Don't run `bundle update`. It's a honorable thought of you but some dependency versions (e.g. pg) are broken 😢 Just use `bundle install` if you introduce new dependencies instead.
- Run `rubocop` and `coffeelint` on your changes to respect the code style guide.
- Add tests for your changes. Feel free to ask for support. We are happy to help you.

* The upper textblock will be removed automatically when you submit your Pull Request *
-->

abeluck committed:
Consistently handle the 'disabled' attr in inputs

As discussed in https://community.zammad.org/t/automating-creation-of-custom-object-attributes/3831/5

This commit ensures that the generic input widgets handle the `disabled`
attribute consistently. Before only the checkbox, select and input widgets respected
this attribute.

I've added it to the following widgets

* tree/searchable select
* textarea
* datetime
* date
* richtext: added appropriate css to hide the attachment link and
            grey-out the div
* checkbox
* radio

Now it is possible to define a screens policy like so:
```
screens: {
"edit"=>{
"ticket.customer"=>{"shown"=>true, "required"=>false},
"ticket.agent"=>{
  "shown"=>true,
  "required"=>false,
  "disabled"=>true # <--- add this
}}}
```

Then the fields will show up on the Agent's sidebar with the customer
submitted values, but the Agent will not be able to change the values.

This is not the same as real permissions, as presumably a manually
crafted API request could circumvent the disabled attribute.

But this is a first step towards that direction, and very useful for us
in prod right now.

thorsteneckel commented:
Hey @abeluck - wow! This was like instantly fast. Would you mind sharing some screenshots on how the elements look when they are disabled?

abeluck commented:
Your response was instantly fast :) I was in the process of creating a screenshot while you typed that :)

![zammad-consistent-disabled](https://user-images.githubusercontent.com/1761767/74224213-b2761d00-4cb8-11ea-8204-81d2ee390aab.png)

"Custom attribute" and "Test_ttset" are both intentionally not disabled.

thorsteneckel commented:
That looks great! Would you also mind to add some tests to cover this behavior? `public/assets/tests/form.js` would be the file. We'd be happy to answer your questions.

abeluck commented:
Sure!

How could I run just those frontend tests? `rake test` seems to run the whole suite.

thorsteneckel commented:
The most convenient way is to start Zammad in `development` mode and navigate your browser to http://localhost:3000/tests_form 🚀

abeluck commented:
Gotcha, thanks.

I'm having trouble finding a test that will capture the whole thing from adding `disabled: true` in the screens to rendering it disabled... so that I can modify it for this usecase :)

Is there an existing test that takes a object attribute params object and renders it? The closest I see are `object manager form 1/2/3` but those don't render the final form the agent sees.. just the admin ui. 

If you could point me to it, then I can base my test on it.

abeluck commented:
I added basic render tests (see new changes) that use the "disabled" attribute in the form params object. This works great!

Along the way I added disabled support for `radiobox`, `richtext`, and `checkbox` just to be consistent, though ObjectManager::Attributes cannot have these types. I intentionally did not add disabled support for `password` or `active` as I don't think it makes sense for those.

If you also want a test that tests all the way from the ObjectManager::Attribute screens hash, then I'll need some help as per my previous comment.

abeluck commented:
Here are screenshots of the new widgets

![zammad-disabled-check](https://user-images.githubusercontent.com/1761767/74229767-4f3db800-4cc3-11ea-94ff-a14523239332.png)
![zammad-disabled-radio](https://user-images.githubusercontent.com/1761767/74229773-51077b80-4cc3-11ea-9529-60e86ccd2438.png)

I didn't like how the richtext edit still looked editable, even though it wasn't, so I added some appropriate css and hid the "select attachment" link.

before css (it looks editable even though it isn't):
![zammad-disabled-contenteditable](https://user-images.githubusercontent.com/1761767/74230789-5ebe0080-4cc5-11ea-98a9-c7be6b128ff4.png)

after css (now it looks properly disabled):
![zammad-disabled-contenteditable2](https://user-images.githubusercontent.com/1761767/74230754-5239a800-4cc5-11ea-9bdf-a1f3279b8c87.png)

thorsteneckel commented:
Damn sweet! @mantas could you please have a look and create a MR to check the impact on CI?

abeluck commented:
@mantas @thorsteneckel  I know a release is imminent, and chance this PR and/or  [Fix deserialization of multiselect value #2932](https://github.com/zammad/zammad/pull/2932) could make it in?

thorsteneckel commented:
@mantas could you please prioritize this PR?

mantas commented:
@thorsteneckel sure, I'll look into this today

mantas commented:
Hi @abeluck ,

Unfortunately this PR won't make it to the upcoming release. While PR itself is great, it brought our attention to a semi-related part of the codebase that may need some love before pulling this one in. We hope to include this in next release.

Sorry for the delay.

abeluck commented:
Understood. Looking forward to the next release then.

abeluck commented:
Now that we're passed the big 3.3 release, any chance we could get this PR looked at again?
